File: challenges-in-python/backtracking/n_queens.py
# Solving the "N-Queens" Challenge
# Challenge Description:
# The N-Queens puzzle is the problem of placing N chess queens on an N×N chessboard 
# so that no two queens threaten each other. Given an integer N, 
# return all distinct solutions to the N-queens puzzle. 
# Each solution contains a distinct board configuration of the queen's placement, 
# where 'Q' and '.' both indicate a queen and an empty space, respectively.

# Approach:
# Use backtracking to explore all possible positions for placing queens one row at a time, 
# applying constraints to ensure no two queens can attack each other.

def solve_n_queens(n):
    def is_valid(board, row, col):
        for i in range(row):
            if board[i] == col or abs(row - i) == abs(col - board[i]): # Check if queens are in the same column or diagonal
                return False
        return True

    def place_queens(board, row):
        if row == n:
            result.append([''.join('Q' if i == col else '.' for i in range(n)) for col in board])
            return
        for col in range(n):
            if is_valid(board, row, col):
                board[row] = col # Place queen
                place_queens(board, row + 1) # Explore next row

    result = []
    place_queens([0] * n, 0)
    return result

# Example Usage
def test_solve_n_queens():
    results = solve_n_queens(4)
    expected = [
        [".Q..","...Q","Q...","..Q."],
        ["..Q.","Q...","...Q",".Q.."]
    ]
    assert len(results) == 2
    assert sorted(results) == sorted(expected)

File: challenges-in-rust/src/backtracking/n_queens.rs
// Solving the "N-Queens" Challenge
// Problem Statement
// The n-queens puzzle is the problem of placing n queens on an n×n chessboard 
// such that no two queens attack each other. 
// Given an integer n, return all distinct solutions to the n-queens puzzle.

// Big O:
// Time complexity: O(n!) | Space complexity: O(n^2)
pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
    let mut board = vec![vec!['.'; n as usize]; n as usize];
    let mut solutions = Vec::new();
    solve(&mut board, 0, &mut solutions);
    solutions
}

// Check if a queen can be placed on board[row][col]
// This function is called when "col" queens are already placed in columns from 0 to col -1.
// So we need to check only left side for attacking queens
fn solve(board: &mut Vec<Vec<char>>, col: usize, solutions: &mut Vec<Vec<String>>) {
    if col == board.len() { 
        solutions.push(board.iter().map(|row| row.iter().collect()).collect());
        return;
    }
    for row in 0..board.len() {
        if is_safe(board, row, col) {
            board[row][col] = 'Q';
            solve(board, col + 1, solutions);
            board[row][col] = '.';
        }
    }
}

fn is_safe(board: &Vec<Vec<char>>, row: usize, col: usize) -> bool {
    for i in 0..col {
        for j in 0..board.len() {
            if board[j][i] == 'Q' && (row == j || row as i32 - col as i32 == j as i32 - i as i32 || row as i32 + col as i32 == j as i32 + i as i32) {
                return false;
            }
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_4_queens() {
        let solutions = solve_n_queens(4);
        assert_eq!(solutions.len(), 2); // There are 2 distinct solutions for 4-queens puzzle
    }

    #[test]
    fn test_1_queens() {
        let solutions = solve_n_queens(1);
        assert_eq!(solutions.len(), 1); // There is 1 distinct solution for 1-queens puzzle
    }
}